zenoh/src/net/routing/interceptor/access_control.rs

```diff
@@ -0,0 +1,337 @@
+                    (false, true) => (None, Some(egress_interceptor)),
+                    (false, false) => (None, None),
+                }
+            }
+            Err(e) => {
+                tracing::error!("[ACCESS LOG]: Failed to get zid with error :{}", e);
+                (None, None)
+            }
+        }
+    }
+
+    fn new_transport_multicast(
+        &self,
+        _transport: &TransportMulticast,
+    ) -> Option<EgressInterceptor> {
+        tracing::debug!("[ACCESS LOG]: Transport Multicast is disabled in interceptor");
+        None
+    }
+
+    fn new_peer_multicast(&self, _transport: &TransportMulticast) -> Option<IngressInterceptor> {
+        tracing::debug!("[ACCESS LOG]: Peer Multicast is disabled in interceptor");
+        None
+    }
+}
+
+impl InterceptorTrait for IngressAclEnforcer {
+    fn compute_keyexpr_cache(&self, key_expr: &KeyExpr<'_>) -> Option<Box<dyn Any + Send + Sync>> {
+        Some(Box::new(key_expr.to_string()))
+    }
+    fn intercept<'a>(
+        &self,
+        ctx: RoutingContext<NetworkMessage>,
+        cache: Option<&Box<dyn Any + Send + Sync>>,
+    ) -> Option<RoutingContext<NetworkMessage>> {
+        let key_expr = cache
+            .and_then(|i| match i.downcast_ref::<String>() {
+                Some(e) => Some(e.as_str()),
+                None => {
+                    tracing::debug!("[ACCESS LOG]: Cache content was not of type String");
+                    None
+                }
+            })
+            .or_else(|| ctx.full_expr())?;
+
+        match &ctx.msg.body {
+            NetworkBody::Push(Push {
+                payload: PushBody::Put(_),
+                ..
+            }) => {
+                if self.action(Action::Put, "Put", key_expr) == Permission::Deny {
+                    return None;
+                }
+            }
+            NetworkBody::Request(Request {
+                payload: RequestBody::Query(_),
+                ..
+            }) => {
+                if self.action(Action::Get, "Get", key_expr) == Permission::Deny {
+                    return None;
+                }
+            }
+            NetworkBody::Declare(Declare {
+                body: DeclareBody::DeclareSubscriber(_),
+                ..
+            }) => {
+                if self.action(Action::DeclareSubscriber, "Declare Subscriber", key_expr)
+                    == Permission::Deny
+                {
+                    return None;
+                }
+            }
+            NetworkBody::Declare(Declare {
+                body: DeclareBody::DeclareQueryable(_),
+                ..
+            }) => {
+                if self.action(Action::DeclareQueryable, "Declare Queryable", key_expr)
+                    == Permission::Deny
+                {
+                    return None;
+                }
+            }
+            _ => {}
+        }
+        Some(ctx)
+    }
+}
+
+impl InterceptorTrait for EgressAclEnforcer {
```

Comment:
Is there any difference between ingress and egress processing?
If not, then it would be better to have one class, why duplicate the code?

Reply:
The idea behind it is to keep the option for allowing separate behavior for ingress and egress in the future